base/tps/src/org/dogtagpki/server/tps/TPSTokendb.java

```diff
@@ -243,25 +243,76 @@ public void tdbAddCertificatesForCUID(String cuid, ArrayList<TPSCertRecord> cert
+                 issuer = _issuer;
+             }
+
+             public String toString() {
+                 return "(cn=" + cn + ", issuerCn=" + issuer + ")";
+             }
+         }
+
+         ArrayList<CnIssuerPair> cnIssuerPairsRemaining = new ArrayList<CnIssuerPair>(certs.size());
+         for(TPSCertRecord cert : certs) {
+             String cn = cert.getId();
+             String issuerCn = cert.getIssuedBy();
+             cnIssuerPairsRemaining.add(new CnIssuerPair(cn, issuerCn));
+         }
+
+        boolean testAddCertsFailure = false;
+	//Contrive a very difficult to reproduce testing scenario
+	CMSEngine engine = CMS.getCMSEngine();
+	try {
+	    EngineConfig configStore = engine.getConfig();
+            // get conn ID
+            String config = "op.enroll." + "testAddCertsToDBFailure";
+            testAddCertsFailure = configStore.getBoolean(config,false);
+        } catch (Exception e) {
+           testAddCertsFailure = false;
+        }
+
         try {
+	    int count = 0;
             for (TPSCertRecord cert : certs) {
                 try {
                     if (!isCertOnToken(cert, cuid)) {
                         logger.debug(method + " adding cert with serial: " + cert.getSerialNumber());
+			// After at least one cert is added correctly, perform the test of a failure
+                        // if so configured.
+                        if(count > 0 && testAddCertsFailure == true) {
+                            throw new Exception("Failed to add certificate to token db, as part of a test of failure condition.");
+                        }
                         tps.certDatabase.addRecord(cert.getId(), cert);
                     } else {
                         // cert already on token
                         logger.debug(method + "retain and skip adding with serial:" + cert.getSerialNumber());
                     }
+		    // Successfully added cert or verified it was already there, so remove
+                    // it from the 'remaining' list
+                    cnIssuerPairsRemaining.removeIf(p -> (p.cn == cert.getId() && p.issuer == cert.getIssuedBy()));
+                    count ++ ;
                 } catch (Exception e) {
                     logger.warn(method + "Exception after isCertOnToken call: "+ e.getMessage(), e);
                     // ignore; go to next;
                 }
             }
         } catch (Exception e) {
             logger.error(method + e.getMessage(), e);
-            // TODO: what if it throws in the middle of the cert list -- some cert records already updated?
-            throw new TPSException(e.getMessage(), e);
+	     String subjectDn = certs.get(0).getSubject();
+             String logMsg = "Failed to add or verify the following certs for [" + subjectDn + "] in the Certificate DB: ";
+             for(CnIssuerPair pair : cnIssuerPairsRemaining) {
+                 logMsg += pair + "; ";
+             }
+
+             throw new TPSException(logMsg);
```

Comment:
same here. Please try to adhere to the practice of adding the content of "method" to the log message to assist admins in pinpointing the location where the message comes from.

Reply:
This is just a continuation of the previous log message, which will have the method prepended already.